File: 06-TREES/PHP/construct-binary-tree-from-preorder-and-inorder-traversal.php
<?php


# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
class Solution
{
    function buildTree(array &$preorder, array $inorder): ?TreeNode
    {
        if (empty($preorder) || empty($inorder)) {
            return null;
        }

        $root_val = array_shift($preorder);
        $root_index = array_search($root_val, $inorder);

        $root = new TreeNode($root_val);
        $root->left = $this->buildTree($preorder, array_slice($inorder, 0, $root_index));
        $root->right = $this->buildTree($preorder, array_slice($inorder, $root_index + 1));

        return $root;
    }
}


$preorder = [3, 4, 6, 1, 2, 7, 5, 8, 10, 9];
$inorder  = [1, 6, 2, 4, 7, 3, 8, 10, 5, 9];
$output   = [3, 4, 5, 6, 7, 8, 9, 1, 2, null, null, null, 10];

print_r((new Solution())->buildTree($preorder, $inorder));

File: 06-TREES/Python/construct-binary-tree-from-preorder-and-inorder-traversal.py
from typing import *
from TreeNode import TreeNode


# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode | None:
        if not preorder or not inorder:
            return None

        root_val = preorder.pop(0)
        root_index = inorder.index(root_val)

        root = TreeNode(root_val)
        root.left = self.buildTree(preorder, inorder[:root_index])
        root.right = self.buildTree(preorder, inorder[root_index + 1:])

        return root


preorder = [3, 9, 20, 15, 7]
inorder  = [9, 3, 15, 20, 7]
Output   = [3, 9, 20, None, None, 15, 7]

TreeNode.fromArray(Output).print()
print()
Solution().buildTree(preorder, inorder).print()